File: 0143-Reorder-List/0143.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: void Do not return anything, modify head in-place instead.
        """
        if head == None or head.next == None:
            return

        pre = head
        lat = head.next
        while lat != None and lat.next != None:
            pre = pre.next
            lat = lat.next.next

        p = pre.next
        pre.next = None
        # reverse
        
        cur = None        
        while p != None:
            q = p.next
            p.next = cur
            cur = p 
            p = q
               
        pre = head
        while pre != None and cur != None:
            tmp = cur.next  
            cur.next = pre.next
            pre.next = cur
            pre = pre.next.next
            cur = tmp 
                   
       

def createList():
    head = ListNode(1)
    cur = head
    for i in range(2, 5):
        cur.next = ListNode(i)
        cur = cur.next
    return head
            
def printList(head):
    cur = head
    while cur != None:
        print(cur.val, '-->', end='')
        cur = cur.next

    print('NULL')

if __name__ == "__main__":
    head = createList()
    printList(head)
    Solution().reorderList(head)
    printList(head)

File: src/0143-Reorder-List/0143.cpp
#include <iostream>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    void reorderList(ListNode* head) 
    {
        if (head == nullptr || head->next == nullptr)
        {
            return;
        }
        ListNode* pre = head;
        ListNode* lat = head->next;
        while (lat != nullptr && lat->next != nullptr)
        {
            pre = pre->next;
            lat = lat->next->next;
        }

        ListNode* p = pre->next;
        pre->next = nullptr;

        ListNode* cur = nullptr;
        while (p != nullptr)
        {
            ListNode* q = p->next;
            p->next = cur;
            cur = p;
            p = q;
        }
        pre = head;
        while (pre != nullptr && cur != nullptr)
        {
            ListNode* tmp = cur->next;
            cur->next = pre->next;
            pre->next = cur;
            pre = pre->next->next;
            cur = tmp;
        }
    }
};
int main()
{
    return 0;
}